packages/api-server/src/schema/squiggle.ts

```diff
@@ -0,0 +1,160 @@
+      resultJSON: Prisma.JsonValue;
+      bindingsJSON: Prisma.JsonValue;
+    }
+);
+
+function runSquiggle(code: string): SquiggleOutput {
+  const MAIN = "main";
+
+  const project = SqProject.create();
+
+  project.setSource(MAIN, code);
+  project.run(MAIN);
+
+  const result = project.getResult(MAIN);
+  const bindings = project.getBindings(MAIN);
+
+  return result.ok
+    ? {
+        isCached: false,
+        isOk: true,
+        resultJSON: squiggleValueToJSON(result.value),
+        bindingsJSON: squiggleValueToJSON(bindings.asValue()),
+      }
+    : {
+        isCached: false,
+        isOk: false,
+        errorString: result.value.toString(),
+      };
+}
+
+const SquiggleOutputObj = builder
+  .interfaceRef<SquiggleOutput>("SquiggleOutput")
+  .implement({
+    fields: (t) => ({
+      isCached: t.exposeBoolean("isCached"),
+    }),
+  });
+
+builder.objectType(
+  builder.objectRef<Extract<SquiggleOutput, { isOk: true }>>(
+    "SquiggleOkOutput"
+  ),
+  {
+    name: "SquiggleOkOutput",
+    interfaces: [SquiggleOutputObj],
+    isTypeOf: (value) => (value as SquiggleOutput).isOk,
+    fields: (t) => ({
+      resultJSON: t.string({
+        resolve(obj) {
+          return JSON.stringify(obj.resultJSON);
+        },
+      }),
+      bindingsJSON: t.string({
+        resolve(obj) {
+          return JSON.stringify(obj.bindingsJSON);
+        },
+      }),
+    }),
+  }
+);
+
+builder.objectType(
+  builder.objectRef<Extract<SquiggleOutput, { isOk: false }>>(
+    "SquiggleErrorOutput"
+  ),
+  {
+    name: "SquiggleErrorOutput",
+    interfaces: [SquiggleOutputObj],
+    isTypeOf: (value) => !(value as SquiggleOutput).isOk,
+    fields: (t) => ({
+      errorString: t.exposeString("errorString"),
+    }),
+  }
+);
+
+builder.queryField("runSquiggle", (t) =>
+  t.field({
+    type: SquiggleOutputObj,
+    args: {
+      code: t.arg.string({ required: true }),
+    },
+    async resolve(_, { code }) {
+      const key = getKey(code);
+
+      const cached = await prisma.squiggleCache.findUnique({
+        where: { id: key },
+      });
+      if (cached) {
+        return {
+          isCached: true,
+          isOk: cached.ok,
+          errorString: cached.error,
+          resultJSON: cached.result,
+          bindingsJSON: cached.bindings,
+        } as unknown as SquiggleOutput; // cache is less strictly typed than SquiggleOutput, so we have to force-cast it
+      }
+      const result = runSquiggle(code);
+      await prisma.squiggleCache.upsert({
+        where: { id: key },
+        create: {
+          id: key,
+          ok: result.isOk,
+          result: result.resultJSON ?? undefined,
+          bindings: result.bindingsJSON ?? undefined,
+          error: result.errorString,
+        },
+        update: {
+          ok: result.isOk,
+          result: result.resultJSON ?? undefined,
+          bindings: result.bindingsJSON ?? undefined,
+          error: result.errorString,
+        },
```

Comment:
This implies that future runs will remove old run results?

Reply:
For now, this code protects us from race conditions (if you ask for a long calculation twice, the second calculation could fail if this code was `insert` and not `upsert`). Outside of race conditions, update part should never happen, until we add cache control. We can revisit this logic when we do that.

In theory, it would be better to de-duplicate long calculations with the identical `key` by holding a mutex, but that's much harder to implement correctly.

Reply:
Oh, interesting. That said, I imagine a better model (to do later) would just not allow this second attempt. Like, we make a record for the run, before it completes, and if a "processing" record exists, we don't allow new runs.